### PR TITLE
DI Vlan module for Brocade Ironware routers

### PR DESCRIPTION
**Pull request by paulquack**
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

ironware_vlan
##### ANSIBLE VERSION

```
ansible 2.5.0 (ironware_vlan 4079a4b5fe) last updated 2017/11/21 21:05:12 (GMT +1100)
  config file = /Users/paul.baker/.ansible.cfg
  configured module search path = [u'/Users/paul.baker/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/paul.baker/ansible/lib/ansible
  executable location = /Users/paul.baker/ansible/bin/ansible
  python version = 2.7.13 (default, Dec 18 2016, 07:03:39) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

##### ADDITIONAL INFORMATION

Unsure of the approach to unit tests around these declarative modules, or how you'd like toaddress different requirements/capabilities of different platforms. For example, I note that VLAN module doesn't dictate whether VLANs are tagged or untagged. My suggestion is that we extend the interfaces parameter in the VLAN to include a "tagging" flag. To me it makes more sense in the VLAN module vs the Interface module, but I haven't had a chance to review the Interface module extensively... Let me know your thoughts! 

```

```

### CODE REVIEW

**Comment by paulquack** [comment]
Addresses #31295 for ironware devices

**Comment by gundalow** [comment]
`net_l2_interface` is where the tagging should be done: http://docs.ansible.com/ansible/latest/net_l2_interface_module.html

**Comment by gundalow** [comment]
Feel free to join `#ansible-network` on Freenode IRC where developers and members of the community are, it may be easier to discuss the other points there.

File: lib/ansible/modules/network/ironware/ironware_vlan.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+    argument_spec.update(ironware_argument_spec)
+
+    required_one_of = [['vlan_id', 'aggregate']]
+    mutually_exclusive = [['vlan_id', 'aggregate']]
+    module = AnsibleModule(argument_spec=argument_spec,
+                           supports_check_mode=True,
+                           required_one_of=required_one_of,
+                           mutually_exclusive=mutually_exclusive)
+
+    check_args(module)
```
Comment (ganeshrn) [inline lib/ansible/modules/network/ironware/ironware_vlan.py]: `check_args()` is a empty function. Is this call required? For other networks platforms `check_args()` is present for legacy reason.

File: lib/ansible/modules/network/ironware/ironware_vlan.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+            continue
+        match = re.match(r'^PORT-VLAN ([^\s,]+)', line)
+        if key and not match:
+            parsed[key] += '\n%s' % line
+        elif match:
+            key = match.group(1)
+            parsed[key] = line
+
+    objs = list()
+    for key, value in parsed.iteritems():
```
Comment (ganeshrn) [inline lib/ansible/modules/network/ironware/ironware_vlan.py]: You can use iteritems as below
```
from ansible.module_utils.six import iteritems
iteritems(parsed)
```

**Comment by ganeshrn** [comment]
@paulquack We plan to add vlan tagging in `net_l2_interface` module.
https://github.com/ansible/ansible/issues/31297

File: lib/ansible/modules/network/ironware/ironware_vlan.py (hunk 1 above)
Comment (paulquack) [reply lib/ansible/modules/network/ironware/ironware_vlan.py]: removed

File: lib/ansible/modules/network/ironware/ironware_vlan.py (hunk 2 above)
Comment (paulquack) [reply lib/ansible/modules/network/ironware/ironware_vlan.py]: updated

**Comment by paulquack** [comment]
I've added some vlan_tagging options to this module as well. In order to specify VLAN membership on the ironware platform you need to say tagged/untagged... so it's a bit tricky to separate the two functions. The module still supports all functionality of the generic DI module, but assumes to add interfaces as tagged.

File: lib/ansible/modules/network/ironware/ironware_vlan.py [hunk 3]
```diff
@@ -75,15 +76,6 @@
     aggregate:
       - vlan_id: 4000
       - {vlan_id: 4001, name: vlan-4001}
-
-- name: Create vlans with a mix of tagged and untagged ports
```
Comment (paulquack) [inline lib/ansible/modules/network/ironware/ironware_vlan.py]: This example resulted in some unpredictable behavior so it has been removed. Suggest vlan could be created then changed to tagged/untagged using the l2interface module...

**Comment by ganeshrn** [comment]
@paulquack IMO it's fine to merge net_vlan and net_l2_vlan functionality if it is a platform dependency. Also, please add either unit or integration tests for this module.

**Review by ganeshrn — COMMENTED** [review]
@paulquack IMO it's fine to merge net_vlan and net_l2_vlan functionality if it is a platform dependency. Also, please add either unit or integration tests for this module.

**Comment by gundalow** [comment]
New functionality freeze for Ansible 2.5 is 15th Jan. If you need assistance with any of this feel free to reach us in `#ansible-network` on Freenode IRC.

**Comment by paulquack** [comment]
Hi Team, Obviously this has missed the deadline for 2.5 and I haven't had a chance to work on this any further. When I do, I'll refactor and submit a new PR.

**Comment by gundalow** [comment]
@paulquack Feel free to join IRC (`#ansible-network`) and we can discuss this more and assist with testing

**Pull request closed without merging**